### ISSUE TITLE
Dynamic Literal Arguments (Forced override suggestions?)

### ISSUE THREAD

**Issue by HielkeMinecraft** [JorelAli/CommandAPI#129]
It would be awesome it this API had an option to dynamically enable/disable literal argument commands.

**Example:**
Say we make a party command. You can create a party with ` /party create`, invite other people with `/party invite <player>` and accept with `/party accept <player>`. You can leave the party with `party leave`, or get info with `party info`

Right now I register every command, and then I run lots of test to make sure people are getting the right result.
For example you can't leave a party when you're not in one. Or join one if you're already in one.

If there was some way the literal argument command could be disabled depending on the context of the command sender, I could only show the `/party create` command if the player is not in a party. Or hide the party accept command if they don't have a pending invite.

I know this is already kind of possible by overriding suggestions with ` overrideSuggestions(Function<CommandSender, String[]> suggestions)`, but the problem is that they are just suggestions. I still have to check if the player is running the correct command.

Honestly, I don't think this is possible because literal arguments are hard coded, but I'm not 100% sure, soooooo

![afbeelding](https://user-images.githubusercontent.com/42968178/90561469-0ea68600-e1a1-11ea-89b0-5cb4676a9e77.png)

It would save me from building in lots of checks, as well as providing a better user experience. 

**Off-topic**
I'm really blown away by how much this project has matured. I love the rewrite you've done and all the arguments you've added. I remember having to create my own hacky scoreboard argument but now it's all just built in! 
Great job!

**Comment by JorelAli** [comment]
To put it simply: yes, what you're proposing is indeed _possible_. Unfortunately, I'm not going to do it. Sorry!

-----

### The idea in general

Honestly, I think this idea is awesome! I agree, current it only provides suggestions and that doesn't really compare to straight up literals. This idea would really showcase the power of Brigadier's command system, but there's too many flaws with it. 

In general, when a server starts up, it constructs this command graph which states what parts of commands come after what (etc. etc.). Then "we stop registering commands". When any of the following events occur, this command graph is sent to the client and that's how they know what they can and cannot run:

- Player joins the game
- Player is made op/deopped
- A player's permissions change 

The CommandAPI already has a handle on this event so it can re-publish the newly constructed command graph [See here](https://github.com/JorelAli/1.13-Command-API/blob/master/1.13CommandAPI/commandapi-core/src/main/java/dev/jorel/commandapi/CommandAPIMain.java#L39), but the big limitation is modifying the graph when the server is running.

Indeed, this is possible! Version 4.0 includes a much more powerful API which lets developers use Brigadier code alongside the CommandAPI's code, which would theoretically allow you to modify the graph once the server is loaded and send this to players as you see fit. As you can imagine, this would _greatly_ increase the complexity of the way the command graph is sent to players (you're basically modifying the graph whilst they're running commands, whilst ALSO tailoring it to specific players).

For this main reason, I don't think I'll be implementing such a system. I believe that the CommandAPI on its own, even with its suggestion system, suffices to meet the needs that the CommandAPI was created for - _to create a non-brigadier "Bukkit-like" API  to let developers create commands that use the Command UI introduced in 1.13_ .

Similarly, certain Minecraft commands are limited by what a user enters - they have suggestions and you can still run commands outside the range of those suggestions. It would be really cool to "lock" suggestions so it can't allow things that aren't included in the suggestions shown to a player, but unfortunately I haven't found a way to do such a thing (or there just straight up isn't a way to do such a thing). Of course, implementing that would also increase the CommandAPI's complexity.

As stated at the beginning, for these reasons I won't implement this - even though it sounds super cool!

-----

### Off topic

Honestly, there are times where I wish I had a forum or something - it would be cool to get a conversation going for this project so I could meet all the people that use the CommandAPI. At the start of the project, I met one of the first developers that used the CommandAPI and I had the luxury to join their community server and get to know all the people there that use my project.

I also like to look back at some of the old documentation (such as the [1.0 documentation](https://github.com/JorelAli/1.13-Command-API/blob/master/docs/v1.0_Documentation.md)) which just shows how much this project has grown. Every single time I feel like "This project can't possibly get any better", I end up adding new features or implementing something wild. Each new update of Minecraft seems to add new features and challenges to overcome (1.6 was really tough!)

Indeed, I feel that this is by far my best project that I've ever done - the countless hours that I've spent in creating the CommandAPI, the design choices that I've had to make, the bugs that I've had to face, and of course it's spectacular documentation. Tomorrow marks the CommandAPI's second birthday - to think that everything started [2 years ago](https://github.com/JorelAli/1.13-Command-API/commit/41baa57b3a9afb070cfb99a71ba64135fdd3c193). I might rewrite the acknowledgements section to celebrate!

**Comment by JorelAli** [comment]
That aside, it might be possible to put a flag on the suggestion system. For example:
```
myArgument.forceOverrideSuggestions(...)
```

And then internally, before the CommandAPI runs the command executor, it checks if the suggestion that the command sender entered was one that was suggested - if not, catch the error before your command executor is invoked. It's "okay" as a solution - it prevents your command executor from being run if the argument that the command sender entered was not one that was suggested, but I don't really like the idea.

Of course, this is the closest thing that I can think of that relates to what you've suggested. I can see if I can implement that if you want?

**Comment by JorelAli** [comment]
Actually, it might be easier to just add it as an option for an argument instead of a `forceOverrideSuggestions()`. For example:
```java
LinkedHashMap<String, Argument> arguments = new LinkedHashMap<>();
arguments.put("int", new IntegerArgument().safeOverrideSuggestions(1, 2, 3).suggestionsOnly());
```

And then, as stated in the previous comment, intercept this internally within the CommandAPI before your command executor is invoked so you (the developer) don't have to deal with this erroneous command

**Comment by HielkeMinecraft** [comment]
I think with force overriding suggestions it is theoretically possible to solve my use case, but it would mean creating one massive command with tons of dynamically forced suggestions. 
It would kind of ruin the clear structure.

Perhaps a better idea is to dynamically show/hide a whole command? Something like this:

```java        
LinkedHashMap<String, Argument> arguments = new LinkedHashMap<>();
        arguments.put("create", new LiteralArgument("create"));
        new CommandAPICommand("party").withArguments(arguments).executesPlayer((p, args) -> {
          //create super cool party
        }).showIf((sender) -> {
           canCreateParty(sender);
        }).register();

boolean canCreateParty( CommandSender sender){
    return today == "Friday";
}
```
This way I don't have to mash all my `/party` commands into one big ugly spaghetti code creation.

I think forced suggestions would be really helpful in other uses cases though! 

### Off topic
I really recommend creating a Discord server. It's a great way to keep contact with your community. I have some experience with setting up a Discord server, I could help if you want? :)
My Discord is Hielke#9427

**Comment by JorelAli** [comment]
"Hiding" parts of a command is handled automatically by permissions. 

Of course, there's absolutely nothing preventing you from adding additional constraints to the permission check! That might be a good way to handle that! I'll have to check whether or not this conflicts with the way permission checks are evaluated (I think they're initially evaluated when a player joins and then it's cached, but it might be re-calculated when a player runs a command).

I think it might be easier to implement "hiding a command" rather than trying to force suggestions!

-----

### Off topic

I'll see what I can do - I've set up a Slack workspace before, how hard could it be to set up a Discord one? Only downside is that the CommandAPI has no visual branding - i.e. no "icon". Never really thought of anything for this project, but it might be a good time to start!

**Comment by HielkeMinecraft** [comment]
Ahh of course permissions can be the solution! I guess I can add/remove permissions dynamically, but it would be better if the API had something like a `showIf()` method. I feel like using permissions is kind of a hacky way of achieving this, and it should be it's own thing.

**Comment by JorelAli** [comment]
Unfortunately, not everything went as planned. Indeed, using predicates to determine if a player is capable of executing a command would work, except for the packet limitation: As stated earlier, a packet is sent when any of these events occur:

- Player joins the game (By the CommandAPI)
- Player is made op/deopped
- A player's permissions change

This process does two things:

- Evaluates all of the commands that a player can run
- Sends this information to the player

For example, I have the following code:

```java
LinkedHashMap<String, Argument> arguments = new LinkedHashMap<>();
arguments.put("test", new LiteralArgument("prize").withRequirement(s -> ((Player) s).getLevel() >= 20));

new CommandAPICommand("hello")
	.withArguments(arguments)
	.executesPlayer((player, args) -> {
		player.getInventory().addItem(new ItemStack(Material.DIAMOND));
	})
	.register();
```

The idea is that if you don't have 20 or more levels, you can't run `hello prize` because it's not valid. If I have 19 levels and try to run the command, it doesn't work. This is to be expected. Now say I kill some mobs or whatever and now have 21 levels. If I run the command, it still doesn't work _because the packet has not been resent_. To fix this, the player would have to re-log on the server so the packet is sent again, which is less than ideal.

-----

Of course, the easy solution is to just send the packet when the check happens! Except this causes a feedback loop - sending the packet when checking the permissions requires the permissions to be checked, which sends the packet which requires the permissions to be checked... - it's a StackOverflowException of packets being sent to the client which crashes the server.

A stupid idea is to constantly send the player packets every _n_ ticks or something in a scheduled repeating task, but that's beyond acceptable.

**Comment by JorelAli** [comment]
Implemented using the `updateRequirements(player);` method:

```java
LinkedHashMap<String, Argument> arguments = new LinkedHashMap<>();
arguments.put("create", new LiteralArgument("create").withRequirement(s -> !s.hasParty()));
arguments.put("partyName", new StringArgument());

new CommandAPICommand("party")
  .withArguments(arguments)
  .executesPlayer((player, args) -> {
    //For example
    player.setParty((String) args[0]);
    CommandAPI.updateRequirements(player);
  })
  .register();
```

This method updates the requirements for a player manually and I'm leaving it up to the developer to determine when is the best time to invoke this method call.

Closing issue as it's been implemented in #130
